README.md
KINDISPERSE 0.10.1
================

-   [1. Introduction](#1-introduction)
-   [2. Installation](#2-installation)
-   [3. The shiny app](#3-the-shiny-app)
    -   [3.1 Running the app](#31-running-the-app)
    -   [3.2 External interface](#32-external-interface)
-   [4. The R package](#4-the-r-package)
    -   [4.1 Simulations and Sampling](#41-simulations-and-sampling)
    -   [4.2 Data Management](#42-data-management)
    -   [4.3 Estimating dispersal](#43-estimating-dispersal)
    -   [4.4 Adapting to a new species:
        *Antechinus*](#44-adapting-to-a-new-species-antechinus)

<!-- README.md is generated from README.Rmd. Please edit that file -->
<!-- badges: start -->
<!-- badges: end -->

The goal of kindisperse is to simulate and estimate close-kin dispersal
kernels.

# 1. Introduction

Dispersal is a key evolutionary process that connects organisms in space
and time. Assessing the dispersal of organisms within an area is an
important component of estimating risks from invasive species, planning
pest management operations, and evaluating conservation strategies for
threatened species.

Assessing the dispersal of small, abundant and short-lived animals such
as insects has traditionally been more difficult than for animals that
can be easily tagged. Responding to this challenge, researchers have
developed various methods based around mark-release-recapture that mark
the organisms with dyes, paint, or chemical tags, before releasing the
individuals and in various ways measuring the number of recaptures.

Such methods suffer the limitations of requiring manipulation of the
same individuals in which dispersal is being assessed, are
labour-intensive when conducted across a large enough area to be
informative, and typically are not estimates of true intergenerational
dispersal (which is measured life-stage to life-stage, e.g. from the egg
of the parent to the egg of its offspring). Such life-stage to
life-stage estimates are important as they are readily interpretable
within established intergenerational analytic frameworks such as
Wright’s neighbourhood size.

Leveraging decreases in sequencing costs, out new method instead
estimates dispersal from the spatial distribution of close-kin. This
method requires that close kin dyads be identified and scored for two
variables: (i) the geographical distance between the two individuals in
the dyad, and (ii) their estimated order of kinship (1st order
e.g. full-sib; 2nd order e.g. half-sib; 3rd order e.g. first cousin).

Close-kin-based dispersal can provide an estimate of the
intergenerational (or parent-offspring) dispersal kernel - a key factor
that connects biological events across the lifespan of an organism with
broader demographic and population-genetic processes such as isolation
by distance. A dispersal kernel is the probability density function
describing the distributions of the locations of dispersed individuals
relative to the source point. Intergenerational dispersal kernels
themselves can be framed in terms of any number of breeding and
dispersal processes, defined by both reference life-stage and number of
generations, and leave their mark in the spatial distribution of various
categories of close kin, which can be treated as samplings from a set of
underlying kernels. Actual kernels vary, but are typically described in
terms of sigma, the second moment of the kernel, also known as its scale
parameter. More complex kernels can also incorporate a parameter for
shape or kurtosis (kappa), representing the fourth moment of the kernel.

In the case of an insect like the mosquito, the most basic
intergenerational kernel, the lifespan or parent-offspring kernel,
reflects all dispersal and breeding processes connecting (e.g.) the
immature (egg, larval, pupal) location of a parent to the immature
location of its offspring. However, this kernel can be combined with
additional breeding, dispersal and sampling events to produce other,
composite dispersal or distribution kernels that contain information
about intergenerational dispersal. For example, the distribution of two
immature full-sibling mosquitoes reflects not a full lifespan of
dispersal, but two ‘draws’ from the component kernel associated with the
mother’s ovipositing behaviour. Were we to sample the same full-sibling
females as ovipositing adults, this would instead represent two draws
from a composite ‘lifespan and additional oviposition’ kernel. Avuncular
larvae, should they exist, would represent draws from related but
distinct intergenerational dispersal kernels - an oviposition kernel,
and a composite ‘oviposition and lifespan’ kernel. The avuncular
distribution kernel would thus reflect a further compositing of these
dispersal events.

There is a rich literature examining the kernels of basic dispersal
events, and analysing them in terms of various kernel functions, whether
Gaussian, exponential, or others with differing properties and shapes,
often reflecting the tendency of dispersal events to be
disproportionately clustered around the source and/or be dispersed at
great distances from the source (i.e. for the kernel to be fat-tailed).
Most of this literature explores dispersal in terms of probability of a
dispersed sample being at a certain radius from the dispersed source. In
the case of close-kin recaptures of e.g. first cousins, we are instead
presented with dispersal events that must be approached in two
dimensions with respect to both radius of dispersal and additionally
angle of dispersal. A successful estimator of intergenerational
dispersal using close-kin recaptures must find strategies to decompose
the extraneous spatial and breeding components affecting the kernels,
and ultimately re-express dispersal in terms of an axial sigma - that
aspect of dispersal which operates within one dimension across a
two-dimensional space. This is the sigma component relied upon by
[Wright (1946)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1209315/)
for isolation by distance , and which is reflected in estimations of
neighbourhood area.

The method we have developed relies upon the fact that different kinship
categories reflect different but related underlying intergenerational
dispersal composites, and uses the relationships between these kinship
distribution kernels to extract information about the core
parent-offspring dispersal kernel that produced the derivative kernels.
For example, the immature distribution kernel of full siblings differs
from the immature distribution of first cousins by a single lifespan:
using an additive variance framework, the first cousin variance that is
not accounted for by subtracting the full sibling variance constitutes
an estimate of the parent-offspring distribution, from which an
intergenerational kernel estimate can be derived. This is because both
immature full-sibling and immature first cousins are ‘phased’ with
respect to the organisms’ life cycle - that is, they are separated by an
integer multiple of parent-offspring dispersal events. It is this
phasing that enables the extraction of a ‘pure’ effective dispersal
estimate, via the additive property of variance. Other examples of
phased relationships include half sibling immatures to half cousin
immatures (one cycle), full sibling immatures to second cousin immatures
(two cycles), or even (for mosquitoes) full sibling immatures to second
cousin ovipositing adults (three cycles).

Further details can be found in [Jasper et
al. (2019)](https://doi.org/10.1111/1755-0998.13043) “A genomic approach
to inferring kinship reveals limited intergenerational dispersal in the
yellow fever mosquito”.

This package supplements these papers by supplying methods for (a)
importing and exporting information about distances & kinship
relationships for pairs of individuals, (b) estimating the axial
distribution (axial sigma for dispersal or position distributions) from
empirical distributions of kin-pairs, and (c) estimating the
intergenerational (parent-offspring) dispersal distribution (axial
sigma) that underlies the distributions of multiple phased kin
categories. This package also implements several simulation tools for
further exploring and testing the properties of intergenerational
dispersal kernels, as well as to assist in designing experiment layouts
& sampling schemes. Finally, for ease of use, the package supplies an
integrated shiny app which also implements the vast majority of package
functionality in a user-friendly interface.

# 2. Installation

You can install the released version of kindisperse from
[CRAN](https://CRAN.R-project.org) with: (in future. this doesn’t
currently exist)

``` r
install.packages("kindisperse")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("moshejasper/kindisperse")
```

Once installed, load the package as follows:

``` r
library(kindisperse)
#> kindisperse 0.10.1
```

# 3. The shiny app

The kindisperse app bundles most of the tools supplied in this package
for ease of use.

## 3.1 Running the app

To run the app, enter the function `run_kindisperse()` and in a moment
the app will appear in a separate window. To close, exit this window, or
alternatively hit the ‘stop’ button or equivalent in RStudio.

## 3.2 External interface

To calculate axial values, etc. of objects within the app, they first
must be passed to the app from the computer or the R package
environment. One option is to save objects to the computer via either
.csv or .kindata formats, then load them using the in-app interface.

Alternatively, objects you have loaded or created in the R package
environment can be passed to the app by first mounting them to the
special `appdata` environment which can be accessed from within the app
via the `Load` tab. Mounted objects must be of class `KinPairData` or
`KinPairSimulation`. To mount an object, use the
`mount_appdata(obj, "nm")` function (unmount with
`unmount_appdata("nm")`). The `appdata` environment can be viewed with
`display_appdata()` and cleared with `reset_appdata()`. Objects mounted
to appdata from within the app can also be retrieved with
`retrieve_appdata()` or `retrieveall_appdata()`.

``` r
fullsibs <- simulate_kindist_composite(nsims = 100, ovisigma = 25, kinship = "FS")
reset_appdata()
mount_appdata(fullsibs, "fullsibs")
display_appdata()
#> <environment: kindisperse_appdata>
#> parent: <environment: namespace:kindisperse>
#> bindings:
#>  * fullsibs: <KnPrSmlt>
fullsibs2 <- retrieve_appdata("fullsibs")
reset_appdata()
```

The app also uses a temporary environment for in-app data handling and
storage. Following a session, objects stored in this space can be
bulk-accessed via the function `retrieve_tempdata()`, and reset via the
function `reset_tempdata()`.

# 4. The R package

Package functions and typical usage are introduced below

## 4.1 Simulations and Sampling

KINDISPERSE is is built to run three types of simulation. The first is a
graphical simulation showing the dispersal of close kin over several
generations. The second and third are both simulations of close kin
dyads, one using a simple PO kernel, the other a composite one. The
package also includes one subsampling function to assist in using these
simulations for field study design.

### 4.1.1 Graphical simulations

This is designed primarily for introducing, exploring, and easily
visualising dispersal concepts. It is packaged in two parallel
functions: the simulation function (`simgraph_data()`) and the
visualisation function(`simgraph_graph()`). A standard example of their
use is shown below, in this case, modeling families including first
cousins with a kernel sigma of 25m, and site dimensions of 250x250m. The
first graph shows the dispersal events leading to first cousins within
five of these families.

``` r
## run graphical simulation
graphdata <- simgraph_data(nsims = 1000, posigma = 25, dims = 250)
simgraph_graph(graphdata, nsim = 5, kinship = "1C")
```

<img src="man/figures/README-graphical_simulation-1.png" width="100%" />

However, the options of both can be tweaked to show other data types,
e.g. a pinwheel graph focused on 1,000 first cousin dyads.

``` r
graphdata <- simgraph_data(nsims = 1000, posigma = 25, dims = 250)
simgraph_graph(graphdata, nsims = 1000, pinwheel = T, kinship = "1C")
```

<img src="man/figures/README-simgraphpin-1.png" width="100%" />

or a histogram of first cousin dyads:

``` r
graphdata <- simgraph_data(nsims = 1000, posigma = 25, dims = 250)
simgraph_graph(graphdata, nsims = 1000, histogram = T, kinship = "1C")
```

<img src="man/figures/README-simgraphhist-1.png" width="100%" />

The simgraph functions are also implemented in the ‘Tutorial’ tab of the
kindisperse app.

### 4.1.2 Kinpair Simulations

These are designed for simulating and testing the impacts of various
dispersal and sampling parameters on a dataset, and for testing and
validating the estimation functions. They return an object of class
`KinPairSimulation`, which supplies a tibble (dataframe) of simulation
results, as well as metadata capturing the simulation parameters.

Three kernel types are supported for the next two simulations at
present: `Gaussian`, `Laplace`, and `vgamma` (variance-gamma). These are
passed to the functions with the `method` parameter. If using vgamma,
also supply its shape parameter with the argument `shape`. Small values
of `shape` correspond to an increasingly leptokurtic kernel - i.e. a
strong central clustering with an increased number of very widely spaced
individuals (long tails).

The simple simulation, `simulate_kindist_simple()`, simulates
intergenerational dispersal for each kin category based on a simple
parent-offspring dispersal sigma, with no attempt to distinguish between
the various breeding and dispersal events across a lifespan. For this
reason, it cannot distinguish between full and half siblings (for
example), as immature full and half siblings have been separated by less
than a lifespan’s worth of dispersal; it would render them as at
distance 0 from their parent (if they were in the adult oviposition
stage, however, they would be rendered as at one lifespans’ dispersal
from parents).

Example usage is shown below:

``` r
simulate_kindist_simple(nsims = 5, sigma = 100, method = "Gaussian", kinship = "PO", lifestage = "immature")
#> KINDISPERSE SIMULATION of KIN PAIRS
#> -----------------------------------
#> simtype:      simple 
#> kerneltype:       Gaussian 
#> kinship:      PO 
#> simdims:      100 100 
#> posigma:      100 
#> lifestage:        immature 
#> 
#> tab
#> # A tibble: 5 x 8
#>   id1   id2   kinship distance    x1     y1     x2     y2
#>   <chr> <chr> <chr>      <dbl> <dbl>  <dbl>  <dbl>  <dbl>
#> 1 1a    1b    PO          51.0 43.1  50.9     20.5   5.20
#> 2 2a    2b    PO         140.   8.84 89.4    -81.8 -17.7 
#> 3 3a    3b    PO         155.  32.2  68.0   -120.   95.5 
#> 4 4a    4b    PO         199.  12.0  29.9   -182.   74.6 
#> 5 5a    5b    PO         100.  35.2   0.874   71.8  94.4 
#> -----------------------------------
```

The composite simulation, `simulate_kindist_composite()`, defines four
smaller dispersal movements which make up the lifestage dispersal
kernel. It distinguishes between full and half siblings, cousins, etc.
and handles immature kin dyads that are separated by less than a
lifespan of dispersal (e.g. immature FS). The four phases are ‘initial’
(handling any dispersal between hatching and breeding), ‘breeding’
(movement of the male across the breeding aspect of the cycle), ‘gravid’
(movement of the female after breeding but before deposition of young),
and ‘oviposition’ (movement made while ovipositing/ bearing young). The
addition of the variances of these four kernels together consitutes the
lifespan dispersal kernel; the relationships between different
categories inform the phase. For example, full-siblings, whether sampled
at oviposition or immature states, differ in hatch position based on the
ovipositing movements of the mother (including e.g. skip oviposition in
the case of some mosquitoes). These categories (and any others
containing a full-sib relationship buried in the pedigree) are thus of
the ‘full-sibling’ or ‘FS’ phase. Half siblings, in mosquitoes (which
this package is modelled on) are expected to be due to having the same
father & separate mothers: the last contribution of the father’s
dispersal is at the breeding stage, so the ‘HS’ phase are differentiated
by the breeding, gravid, & oviposition phases, but share in common the
initial phase. The parent-offspring ‘PO’ phase, on the other hand, share
all (or none) of the component dispersal distributions.

An example composite simulation is demostrated below:

``` r
simulate_kindist_composite(nsims = 5, initsigma = 50, breedsigma = 30, gravsigma = 50, ovisigma = 10, method = "Laplace", kinship = "H1C", lifestage = "ovipositional")
#> KINDISPERSE SIMULATION of KIN PAIRS
#> -----------------------------------
#> simtype:      composite 
#> kerneltype:       Laplace 
#> kinship:      H1C 
#> simdims:      100 100 
#> initsigma         50 
#> breedsigma        30 
#> gravsigma         50 
#> ovisigma      10 
#> lifestage:        ovipositional 
#> 
#> tab
#> # A tibble: 5 x 8
#>   id1   id2   kinship distance     x1    y1    x2     y2
#>   <chr> <chr> <chr>      <dbl>  <dbl> <dbl> <dbl>  <dbl>
#> 1 1a    1b    H1C         406. -78.7   89.1  323.   25.6
#> 2 2a    2b    H1C         116. 153.    60.2  148.  176. 
#> 3 3a    3b    H1C         133.  83.8  -59.3  175. -156. 
#> 4 4a    4b    H1C         354.  25.7   22.1 -165. -276. 
#> 5 5a    5b    H1C         167.   8.99 150.   176.  141. 
#> -----------------------------------
```

Finally, a custom simulation function, `simulate_kindist_custom()`
enables the simulation of dispersal in organisms with breeding cycles
different to the original mosquito species this package was modeled for.
These simulations take a model object (of class `DispersalModel`,
generated with the function `dispersal_model()`) which contains detailed
information about breeding phases, the full sibling (FS) & half sibling
(HS) branch point, the sampling point, and further optional parameters
defining the accessible breeding cycle more carefully. We illustrate
this functions’s use in the context of implementing kindisperse in a new
species subsequently, but an initial example is given here. First, we
generate a custom dispersal model:

``` r
dmodel <- dispersal_model(juvenile = 50, breeding = 40, gestation = 30, .FS = "juvenile", .HS = "breeding", .sampling_stage = "gestation")
dmodel
#> KINDISPERSE INTERGENERATIONAL DISPERSAL MODEL
#> ---------------------------------------------
#> stage:        juvenile   breeding    gestation 
#> dispersal:    50 40  30 
#> 
#> FS branch:    juvenile 
#> HS branch:    breeding 
#> sampling stage:   gestation 
#> cycle:        0 0 
#> ---------------------------------------------
```

Next, we use this model to simulate dispersal in our organism:

``` r
simulate_kindist_custom(nsims = 5, model = dmodel, kinship = "PO")
#> KINDISPERSE SIMULATION of KIN PAIRS
#> -----------------------------------
#> simtype:      custom 
#> kerneltype:       Gaussian 
#> kinship:      PO 
#> simdims:      100 100 
#> juvenile              50 
#> breeding              40 
#> gestation             30 
#> cycle:            0 0 
#> lifestage:        gestation 
#> 
#> tab
#> # A tibble: 5 x 8
#>   id1   id2   kinship distance    x1    y1     x2     y2

NEWS.md
# kindisperse 0.10.1
Expanded the `DispersalModel` object & functions to include optional `breeding_stage` & `visible_stage` parameters to further define dispersal.
Altered `simulate_kindist_custom()` & `axials_standard()` etc. functions to explicitly take into account new aspects of the dispersal model, especially `-1` `cycle` numbers (which allow sampling before the 'official' beginning of a lifespan  (often defined as first dispersal). 
Added `cycle` & `override` parameters to `axials_standard` & `axpermute_standard` functions. 
Altered simulation functions & `KinPairSimulation` object to pass a model to the simulation object. 
Added new tutorials to readme & vignettes for applying results to other species.
Vastly improved documentation of almost all functions

# kindisperse 0.10.0
Implemented the 'DispersalModel' class and 'dispersal_model()' constructor function to enable descriptions of non-mosquito breeding & dispersal cycles. 
Implemented the 'simulate_kindist_custom()' function to enable simulation of custom breeding, dispersal & sampling cycled defined by 'DispersalModel' objects. 
Implemented the 'cycle' parameter in the above model object & simulation functions as well as 'axials_standard()' and 'axpermute_standard()' functions to enable the treatment of individuals collected at different (phased) stages of maturity. 
Adjusted various class & function definitions to make compatible with custom dispersal model objects & simulations (e.g. KinPairData, KinPairSimulation, df_to_kinpair())
Added generics & methods for accessing DispersalModel class objects (& the 'breeding_cycle' method for the KinPairData class)


# kindisperse 0.9.2
Changed all occurrences of the 'gamma' kernelshape/method parameter value to 'vgamma' to make it clear we are using a variance-gamma distribution, not a gamma distribution. 
(effects simulation functions)
Removed unused dispersal kernels (simulation functions)
Changed simulation shape parameter default from 1 to 0.5. 

# kindisperse 0.9.1

* Added a `NEWS.md` file to track changes to the package.
* Fixed several display and layout issues in Shiny app
* Changed parameter of 'elongate' function from 'shape' to 'aspect' to make clear we are manipulating aspect ratios of a rectangular area. 

# kindisperse 0.9.0

* Permutation functions now output mean instead of median
* Changed all 'dsigma' parameters & references to 'posigma'
* Changed simulation 'method' parameter input from 'Gamma' to 'gamma'
* Changed 'lifestage' parameter values from 'larva' to 'immature' and from 'oviposition' to 'ovipositional'
* Enable package function messages to be suppressed
* Updated app tutorial and documentation (including vignettes)
* Updated KinPairSimulation '<-sampledim' method
* Added Table of Contents to README

# kindisperse 0.8.0

* Added Mentari Court field data for example calculations
* Added 2d rectangular dimensions to simulate & sample functions
* Changed the way 'sample_kindist' deals with dimensions
* Added helper function 'elongate' for dimension handling
* Changed 'composite' default parameter value from 2 to 1 in function 'axpermute'
* Changed 'juvsigma' parameter & references to 'initsigma' in line with usage in paper
* included rgamma function
* Improved simgraph commands (simgraph_data, simgraph_graph)
* Cleaned up app
* Removed unneeded class methods (to_tibble). Fixed kernelshape method
* Improved package load
* Updated app interface
* Updated documentation (CRAN compatibility)

# kindisperse 0.7.0

* Added options to import data from files within app
* Updated readme, package startup & vignettes
* Added gamma kernel with shape parameter to simulations

# kindisperse 0.6.0

* Changed 'category' parameter in all functions and classed to 'kinship' parameter
* Added file writing and reading functions (write_kindata, read_kindata, kinpair_to_tsv, tsv_to_kinpair, kinpair_to_csv, csv_to_kinpair, kinpair_to_tibble)
* Added appdata interface environment & functions to access it (mount_appdata, unmount_appdata, display_appdata, reset_appdata, retrive_appdata)
* Enable loading of appdata into app (functions and interface)
* Improved KinPairSimulation object print function

# kindisperse 0.5.0

* Replaced file-based memory storage with environment storage in app
* Added 'distances' method for KinPairData & KinPairSimulation classes
* Added functions to test class membership
* Updated axial functions to take KinPairData &c classes as input
* Limit sample size in axial permuation functions under 'std' option
* Updated class definitions
* Added class constructor functions
* Added class import functions
* Updated dependencies

# kindisperse 0.4.0

* Introduced KinPairData & KinpairSimulation classes
* Added and updated class methods & documentation
* Updated app with axial functions
* Fixed bug in 'axials_decompose' function

# kindisperse 0.3.0

* improved NAMESPACE
* add axial function
* document many functions

# kindisperse 0.2.0

* added temp folder & contents to fix app bug
* removed broken dependencies

# kindisperse 0.1.0 & prior

* initial commit


R/KinPairData.R
#' @include DispersalModel.R
NULL

methods::setOldClass(c("tbl_df", "tbl", "data.frame"))

#' Formal class KinPairData
#'
#' The class \code{KinPairData} is a formal (S4) class for storing kinship and lifespan dispersal
#' information concerning kin pairs. It is the base class on which the \code{KinPairSimulation} class is built.
#' The \code{KinPairData} class is used to store information about the spatial distribution of kin dyads for use
#' in calculating axial sigmas of intergenerational dispersal as initially implemented in Jasper et al. 2019
#' (\doi{https://doi.org/10.1111/1755-0998.13043}).
#'
#' This class is essentially wrapped around the \code{tbl_df} class but with (a) expectations around certain columns
#' that must be present (\code{id1, id2, kinship}, & \code{distance} - three 'character' & one 'numeric' column), as
#' well as (b) additional attributes (\code{kinship}, \code{lifestage}, & \code{cycle}) characterizing the close-kin
#' dyads being stored.These attributes, as well as the embedded vector of distances, can be accessed with the methods
#' \code{\link{kinship}}, \code{\link{lifestage}}, \code{\link{breeding_cycle}} and \code{\link{distances}}.
#'
#' Objects from this class are returned from the \code{\link{df_to_kinpair}} and \code{\link{csv_to_kinpair}} functions
#' (& related), and are directly constructed with the namesake \code{KinPairData()} function.
#' They can be passed to the \code{\link{sample_kindist}} function for filtering and subsampling, and to
#' axial functions (including \code{\link{axials_standard}} and \code{\link{axpermute_standard}}) for estimation of
#' axial dispersal.
#' @slot kinship character - one of PO, FS, HS, AV, HAV, GG, 1C, H1C, GAV, HGAV, 1C1, H1C1, GGG, 2C, and H2C.
#' @slot lifestage character - lifestage at sampling - either 'immature', 'ovipositional' or a stage
#' corresponding to a \code{DispersalModel} custom stage
#' @slot cycle non-negative integer or vector of two such integers - Represents
#' the number of complete breeding cycles each  individual has undergone before the sampling point, where the time between
#' birth and first reproduction is coded as '0', that between first and second reproduction '1', etc. (default 0). If the first individual
#' was sampled as a juvenile & the second as an adult of equivalent stage, the vector c(0, 1) would be used. In most situations, the default will be appropriate
#' @slot tab tbl_df. - tibble of dispersal values
#' @return returns object of class \code{KinPairData}
#' @export
#' @family kdclasses
#'
KinPairData <- setClass("KinPairData",
  slots = list(kinship = "character", lifestage = "character", cycle = "numeric", tab = "tbl_df")
)


######### GENERICS and METHODS#############


#' Access or assign \code{kinship} category of \code{\link{KinPairData}} class objects
#'
#' @param x object with relevant method
#'
#' @return returns \code{character} kinship category of object or \code{KinPairData} object with modified kinship category
#' @export
#' @family kpdmethods
#'
setGeneric("kinship", function(x) standardGeneric("kinship"))
#'
#' @rdname kinship
#' @param x object with relevant method
#' @param value new value to assign
#'
#' @export
#'
#'
setGeneric("kinship<-", function(x, value) standardGeneric("kinship<-"))
#' Access or assign \code{lifestage} category of \code{\link{KinPairData}} class objects
#'
#' @param x object with relevant method
#' @return returns \code{character} lifestage of object or \code{KinPairData} object with modified lifestage
#' @export
#' @family kpdmethods
#'
#'
setGeneric("lifestage", function(x) standardGeneric("lifestage"))
#'
#' @rdname lifestage
#' @param x object with relevant method
#' @param value new value to assign
#'
#' @export
#'
#'
setGeneric("lifestage<-", function(x, value) standardGeneric("lifestage<-"))

#' Access or assign \code{distances} category of \code{\link{KinPairData}} class objects
#'
#' @param x Object of Class KinPairData
#'
#' @return Returns a numeric vector of kin separation distances
#' @export
#' @family kpdmethods
#'
#'
setGeneric("distances", function(x) standardGeneric("distances"))




#' @describeIn distances
#'
#' @param KinPairData object of class \code{KinPairData}
#' @param x object of class KinPairData
#'
#' @export
#'
setMethod("distances", "KinPairData", function(x) x@tab$distance)

#' @describeIn kinship
#'
#' @param KinPairData object of class \code{KinPairData}
#'
#' @export
#'
setMethod("kinship", "KinPairData", function(x) x@kinship)

#' @describeIn kinship
#'
#' @param KinPairData object of class \code{KinPairData}
#' @param x object of class \code{KinPairData}
#' @param value value to assign to slot
#'
#' @export
#'
setMethod("kinship<-", "KinPairData", function(x, value) {
  x@kinship <- value
  validObject(x)
  x
})


#' @describeIn lifestage
#'
#' @param KinPairData object of class \code{KinPairData}
#'
#' @export
#'
setMethod("lifestage", "KinPairData", function(x) x@lifestage)

#' @describeIn lifestage
#'
#' @param KinPairData object of class \code{KinPairData}
#'
#' @export
#'
setMethod("lifestage<-", "KinPairData", function(x, value) {
  x@lifestage <- value
  validObject(x)
  x
})

#' @describeIn sampling_stage
#'
#' @param KinPairData object of class \code{KinPairData}
#'
#' @export
#'
#'
setMethod("sampling_stage", "KinPairData", function(x) x@lifestage)

#' @describeIn breeding_cycle
#'
#' @param KinPairData object of class \code{KinPairData}
#'
#' @export
setMethod("breeding_cycle", "KinPairData", function(x) x@cycle)


#' @describeIn KinPairData standard print method
#'
#' @param KinPairData object of class KinPairData
#' @param object an object of class KinpairData
#'
#' @return No return value, called for side effects
#'
#' @export
#'
setMethod(
  "show",
  "KinPairData",
  function(object) {
    cat("KINDISPERSE RECORD OF KIN PAIRS\n")
    cat("-------------------------------\n")
    cat("kinship:\t\t", object@kinship, "\n")
    cat("lifestage:\t\t", object@lifestage, "\n")
    cat("cycle:\t\t\t", object@cycle, "\n")
    cat("\ntab\n")
    print(object@tab)
    cat("-------------------------------")
  }
)

# Constructor method of KinPairData

#' @describeIn KinPairData initialize method
#'
#' @param KinPairData object of class KinPairData
#' @param .Object the KinPairData object to be constructed
#' @param data  data about kinship to be used to construct object (tibble, data.frame, or numeric vector of distances)
#' @param kinship character. Kinship category value for object. - one of PO, FS, HS, AV, HAV, GG, 1C, H1C, GAV, HGAV, 1C1, H1C1, GGG, 2C, and H2C.
#' @param lifestage character. Lifestage value for object. - one of 'immature', 'ovipositional' or 'unknown'
#' @param cycle non-negative integer or vector of two such integers - Represents
#' the number of complete breeding cycles each simulated individual has undergone before the sampling point, where the time between
#' birth and first reproduction is coded as '0', that between first and second reproduction '1', etc. (default 0). If the first individual
#' was sampled as a juvenile & the second as an adult of equivalent stage, the vector c(0, 1) would be used. In most situations, defualt will be appropriate
#' @param ... additional argument to pass to downstream functions in future
#'
#' @export
#'
#' @return Returns an object of class \code{KinPairData}
#'
setMethod(
  "initialize", "KinPairData",
  function(.Object,
           data = NULL,
           kinship = NULL,
           lifestage = NULL,
           cycle = NULL,
           ...) {
    if (!is.null(kinship)) {
      .Object@kinship <- kinship
    }
    else {
      .Object@kinship <- "UN"
    }
    if (!is.null(lifestage)) {
      .Object@lifestage <- lifestage
    }
    else {
      .Object@lifestage <- "unknown"
    }
    if (!is.null(cycle)) .Object@cycle <- cycle else .Object@cycle <- 0
    if (!is.null(data)) {
      if (is.data.frame(data) & !is_tibble(data)) {
        data <- as_tibble(data)
      }
      if (is_tibble(data)) {
        if (ncol(data) == 1) {
          data <- data[[1]]
        }
      }
      if (is_tibble(data)) {
        if (!"distance" %in% colnames(data)) {
          if (!("x1" %in% colnames(data) & "y1" %in% colnames(data) & "x2" %in% colnames(data) & "y2" %in% colnames(data))) {
            stop("Unable to determine kin distances!")
          }
          else {
            data <- mutate(data, distance = sqrt((.data$x1 - .data$x2)^2 + (.data$y1 - .data$y2)^2))
          }
        }
        if (!"kinship" %in% colnames(data)) {
          data <- mutate(data, kinship = .Object@kinship)
        }
        if (!"id1" %in% colnames(data)) {
          data <- add_column(data, id1 = paste0(1:nrow(data), "a"))
        }
        if (!"id2" %in% colnames(data)) {
          data <- add_column(data, id2 = paste0(1:nrow(data), "b"))
        }
        data <- select(data, .data$id1, .data$id2, .data$kinship, .data$distance, everything())
        .Object@tab <- data
      }
      else { # check if just distances included
        if (is.numeric(data)) {
          message("Note: numeric vector interpreted as kin distances")
          data <- tibble(id1 = paste0(1:length(data), "a"), id2 = paste0(1:length(data), "b"), kinship = .Object@kinship, distance = data)
          .Object@tab <- data
        }
      }
    }
    else {
      .Object@tab <- tibble(id1 = "a", id2 = "b", kinship = "UN", distance = 0, .rows = 0)
    }
    validObject(.Object)
    return(.Object)
  }
)

#' Make new KinPairData object
#'
#' @param data tlb_df. Tibble of kinpair distances
#' @param kinship character. - one of PO, FS, HS, AV, HAV, GG, 1C, H1C, GAV, HGAV, 1C1, H1C1, GGG, 2C, H2C & UN.
#' @param lifestage character. - one of 'unknown', 'immature' or 'ovipositional', or alternatively a custom
#' stage that corresponds to a dispersal stage contained in a \code{DispersalModel} object.
#' @param cycle non-negative integer of length one or two (here, 1 is equivalent to c(1, 1)). Represents
#' the number of complete breeding cycles each individual has undergone before the sampling point, where the time between
#' birth and first reproduction is coded as '0', that between first and second reproduction '1', etc. (default 0). If the first individual
#' was sampled as a juvenile & the second as an adult of equivalent stage, the vector c(0, 1) would be used.
#' In most situations, the default will be appropriate
#'
#' @return returns an object of class \code{KinPairData}
#' @export
#'
#' @examples
#' kin_pair_data()
kin_pair_data <- function(data = NULL, kinship = NULL, lifestage = NULL, cycle = NULL) {
  new("KinPairData", data = data, kinship = kinship, lifestage = lifestage)
}


setValidity("KinPairData", function(object) {
  if (!object@kinship %in% c("UN", "PO", "GG", "GGG", "FS", "AV", "GAV", "1C", "1C1", "2C", "HS", "HAV", "HGAV", "H1C", "H1C1", "H2C")) {
    "@kinship must be one of UN PO GG GGG FS AV GAV 1C 1C1 2C HS HAV HGAV H1C H1C1 H2C"
  }
  #else if (!object@lifestage %in% c("unknown", "immature", "ovipositional")) {
  #  "@lifestage must currently be set to 'unknown', 'immature', or 'ovipositional'"
  #}
  else {
    TRUE
  }
})

#' Check if object is of class KinPairData
#'
#' @param x object to be checked
#'
#' @return Returns TRUE if of class \code{KinPairData}, FALSE if not.
#' @export
#'
#'
is.KinPairData <- function(x) {
  "KinPairData" %in% is(x)
}
